src/common/debug.rs

```diff
@@ -0,0 +1,90 @@
+use std::sync::Arc;
+
+use parking_lot::Mutex;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use crate::common::pyroscope_state::pyro::PyroscopeState;
+use crate::settings::Settings;
+
+#[derive(Serialize, JsonSchema, Debug, Deserialize, Clone)]
+pub struct PyroscopeConfig {
+    pub url: String,
+    pub identifier: String,
+    pub user: Option<String>,
+    pub password: Option<String>,
+    pub sampling_rate: Option<u32>,
+}
+
+#[derive(Default, Debug, Serialize, JsonSchema, Deserialize, Clone)]
+pub struct DebugConfig {
+    pub pyroscope: Option<PyroscopeConfig>,
+}
+
+#[derive(Debug, Serialize, JsonSchema, Deserialize, Clone)]
+#[serde(rename_all = "snake_case")]
+pub enum DebugConfigPatch {
+    Pyroscope(Option<PyroscopeConfig>),
+}
+
+pub struct DebugState {
+    pub pyroscope: Arc<Mutex<Option<PyroscopeState>>>,
+}
+
+impl DebugState {
+    pub fn from_settings(settings: &Settings) -> Self {
+        let pyroscope_config = settings.debug.pyroscope.clone();
+        Self {
+            pyroscope: Arc::new(Mutex::new(PyroscopeState::from_config(pyroscope_config))),
+        }
+    }
+
+    pub fn get_config(&self) -> DebugConfig {
+        let pyroscope_config = {
+            #[cfg(target_os = "linux")]
+            {
+                let pyroscope_state_guard = self.pyroscope.lock();
+                pyroscope_state_guard.as_ref().map(|s| s.config.clone())
+            }
+            #[cfg(not(target_os = "linux"))]
+            {
+                None
+            }
+        };
+
+        DebugConfig {
+            pyroscope: pyroscope_config,
+        }
+    }
+
+    pub fn apply_config_patch(&self, patch: DebugConfigPatch) -> bool {
+        #[cfg(target_os = "linux")]
+        {
+            match patch {
+                DebugConfigPatch::Pyroscope(new_config) => {
+                    if let Some(mut pyroscope_state) = self.pyroscope.lock().take() {
+                        let stopped = pyroscope_state.stop_agent();
+                        if !stopped {
+                            return false;
+                        }
+                    }
+
+                    match new_config {
+                        Some(new_config) => {
+                            let mut pyroscope_guard = self.pyroscope.lock();
+                            *pyroscope_guard = PyroscopeState::from_config(Some(new_config));
+                            true
+                        }
+                        None => true,
+                    }
```

Comment:
We must keep the lock here across the take-reinsert to prevent two pyroscope agents from being instantiated at the same time.

Also, is there a strict need for us to explicitly stop the existing instance with every change?

Reply:
> must keep the lock here

Good catch. Done

> is there a strict need for us to explicitly stop the existing instance with every change?
Afaik, yes. Especially given that [their code](https://github.com/grafana/pyroscope-rs/blob/main/src/pyroscope.rs#L662-L711), uses `clone` on the config that we pass and runs it in a separate thread.